### ISSUE TITLE
Sunset automation runs multiple times

### ISSUE THREAD

**Issue by bachya** [home-assistant/appdaemon#129]
I have the following app to turn on lights 30 minutes before sunset (this is in `lights.py`):

```Python
import appdaemon.appapi as appapi


class EveningLights(appapi.AppDaemon):
    """Define a class to represent the app."""

    def initialize(self):
        """Initialize."""
        self.run_at_sunset(self.sunset_cb, offset=-60 * 30)

    def sunset_cb(self, kwargs):
        """Turn on the living room lights just before sunset."""
        self.log('30 minutes before sunset triggered; turning on lights')
        self.turn_on('group.living_room_lights')
```
The app ran tonight, but according to the logs, it seemed to run three times:

```
2017-08-26 19:10:09.020369 INFO evening_lights: 30 minutes before sunset triggered; turning on lights
2017-08-26 19:18:14.024341 INFO evening_lights: 30 minutes before sunset triggered; turning on lights
2017-08-26 19:34:07.026033 INFO evening_lights: 30 minutes before sunset triggered; turning on lights
```

Not sure why?

**Comment by acockburn** [comment]
IS there anything else in the logs? Can I see the whole log from startup please?

**Comment by bachya** [comment]
@acockburn: You got it: as soon as it runs this evening, I'll send along. Thanks!

**Comment by acockburn** [comment]
OK, are you setting a timezone in your config? I just saw something that might be related to this.

**Comment by bachya** [comment]
I'm not. Here's my non-app-specific config:

```
AppDaemon:
  logfile: STDOUT
  errorfile: STDERR
  threads: 10
  api_port: 5051
  api_key: <APPDAEMON API KEY>
HASS:
  ha_url: <HOME ASSISTANT URL>
  ha_key: <HOME ASSISTANT KEY>
```

**Comment by acockburn** [comment]
OK, Can you add a line in your app to run `get_scheduler_entries()` and log it, at the end of `initialize()` and also in the callback then get me the output. Something like:

`self.log(self.get_scheduler_entries()))`

**Comment by bachya** [comment]
Will do!

**Comment by bachya** [comment]
That seems to throw an exception:

```
2017-09-19 12:39:40.833440 INFO Loading Module: /config/apps/lights.py
2017-09-19 12:39:40.836567 INFO Loading Object living_room_lights using class LivingRoomLights from module lights
2017-09-19 12:39:40.838562 WARNING ---------------------------------------------------------
2017-09-19 12:39:40.839182 WARNING Unexpected error during loading of living_room_lights:
2017-09-19 12:39:40.839805 WARNING ---------------------------------------------------------
2017-09-19 12:39:40.845366 WARNING Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/appdaemon/appdaemon.py", line 924, in read_app
    init_object(name, class_name, module_name, config[name])
  File "/usr/lib/python3.5/site-packages/appdaemon/appdaemon.py", line 598, in init_object
    conf.objects[name]["object"].initialize()
  File "/config/apps/lights.py", line 36, in initialize
    self.log(self.get_scheduler_entries())
  File "/usr/lib/python3.5/site-packages/appdaemon/appapi.py", line 172, in log
    msg = self._sub_stack(msg)
  File "/usr/lib/python3.5/site-packages/appdaemon/appapi.py", line 150, in _sub_stack
    msg = json.dumps(msg)
  File "/usr/lib/python3.5/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.5/json/encoder.py", line 198, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.5/json/encoder.py", line 256, in iterencode
    return _iterencode(o, 0)
TypeError: keys must be a string

2017-09-19 12:39:40.845865 WARNING ---------------------------------------------------------
```

For reference, here's what `lights.py` looks like:

```python
"""Define automations for lighting."""

import appdaemon.appapi as appapi


class LivingRoomLights(appapi.AppDaemon):
    """Define a class to represent the app."""

    ENABLED_SWITCH = 'input_boolean.app_lights'

    def initialize(self):
        """Initialize."""
        self.proximity = self.get_app('proximity')

        self.run_at_sunset(
            self.sunset_cb,
            offset=-60 * 30,
            constrain_input_boolean=self.ENABLED_SWITCH)

        self.log(self.get_scheduler_entries())

    def sunset_cb(self, kwargs):
        """Turn on the living room lights just before sunset."""
        self.log(self.get_scheduler_entries())
        if self.proximity.anyone_home():
            self.log('30 minutes before sunset triggered; turning on lights')
            self.turn_on('group.living_room_lights')
```

**Comment by acockburn** [comment]
Ahh you're running an older version that doesn't support logging of dictionaries directly - I fixed that a few releases ago.

Try this:

```python
self.log(str(self.get_scheduler_entries()))
```

**Comment by bachya** [comment]
I'll upgrade! Shame on me for being behind. :)

**Comment by bachya** [comment]
Sorry for the delay – here's the log:

```
2017-09-22 13:41:32.526349 INFO AppDaemon Version 2.1.9 starting
2017-09-22 13:41:32.526985 INFO Configuration read from: /config/appdaemon.yaml
2017-09-22 15:39:13.937250 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:18.961100 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:23.983811 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:29.010511 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:34.032604 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:39.054706 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:44.076412 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:49.098113 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:54.120553 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:39:59.143449 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:40:04.169581 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:40:09.192513 WARNING Disconnected from Home Assistant, retrying in 5 seconds
2017-09-22 15:40:14.347482 INFO Got initial state
2017-09-22 15:40:14.363025 INFO Reloading Module: /config/apps/proximity.py
2017-09-22 15:40:14.368508 INFO Loading Object proximity using class Proximity from module proximity
2017-09-22 15:40:14.369844 INFO Reloading Module: /config/apps/const.py
2017-09-22 15:40:14.372286 INFO Reloading Module: /config/apps/modes.py
2017-09-22 15:40:14.375230 INFO Loading Object vacation_mode using class VacationMode from module modes
2017-09-22 15:40:14.376120 INFO Loading Object guest_mode using class GuestMode from module modes
2017-09-22 15:40:14.377263 INFO Reloading Module: /config/apps/util.py
2017-09-22 15:40:14.381198 INFO Loading Object util using class Util from module util
2017-09-22 15:40:14.382152 INFO Reloading Module: /config/apps/hass.py
2017-09-22 15:40:14.385528 INFO Loading Object in_blackout using class InBlackout from module hass
2017-09-22 15:40:14.459518 INFO Reloading Module: /config/apps/notifier.py
2017-09-22 15:40:14.464146 INFO Loading Object notifier using class Notifier from module notifier
2017-09-22 15:40:14.465682 INFO Reloading Module: /config/apps/lights.py
2017-09-22 15:40:14.468991 INFO Loading Object living_room_lights using class LivingRoomLights from module lights
2017-09-22 15:40:14.471490 INFO living_room_lights: {'in_blackout': {UUID('7dca5394-0c63-4fc9-beef-09ce3e52db38'): {'kwargs': 1506171600, 'offset': 1506171600, 'name': 'in_blackout', 'basetime': 1506171600, 'type': None, 'repeat': 1506171600, 'interval': 1506171600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506171600}, UUID('136c2985-85a2-4d2e-b371-c9203589cd45'): {'kwargs': 1506135600, 'offset': 1506135600, 'name': 'in_blackout', 'basetime': 1506135600, 'type': None, 'repeat': 1506135600, 'interval': 1506135600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506135600}}, 'security': {UUID('ed697082-9204-491e-a9d0-e35e00b4aed7'): {'kwargs': 1506116651, 'offset': 1506116651, 'name': 'security', 'basetime': 1506116651, 'type': None, 'repeat': 1506116651, 'interval': 1506116651, 'callback': <bound method Security.absent_and_insecure_cb of <security.Security object at 0x75892d10>>, 'timestamp': 1506116651}, UUID('58611465-e747-43d2-974a-db25ae317360'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'security', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method Security.lock_up_at_midnight_cb of <security.Security object at 0x75892d10>>, 'timestamp': 1506146400}}, 'living_room_lights': {UUID('7cc47537-5c49-40e5-bb03-4d53b81ca64c'): {'kwargs': 1506128204, 'offset': 1506128204, 'name': 'living_room_lights', 'basetime': 1506128204, 'type': 'next_setting', 'repeat': 1506128204, 'interval': 1506128204, 'callback': <bound method LivingRoomLights.sunset_cb of <lights.LivingRoomLights object at 0x75842c30>>, 'timestamp': 1506126404}}, 'systems_maintenance': {UUID('601ce8f0-b1c2-4ba7-95c0-46480308ff5e'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'systems_maintenance', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method SystemsMaintenance.midnight_cb of <systems.SystemsMaintenance object at 0x75842050>>, 'timestamp': 1506146400}}}
2017-09-22 15:40:14.472814 INFO Reloading Module: /config/apps/security.py
2017-09-22 15:40:14.475946 INFO Loading Object security using class Security from module security
2017-09-22 15:40:14.478447 INFO Reloading Module: /config/apps/systems.py
2017-09-22 15:40:14.481831 INFO Loading Object systems_notifications using class SystemsNotifications from module systems
2017-09-22 15:40:14.483339 INFO Loading Object systems_maintenance using class SystemsMaintenance from module systems
2017-09-22 15:40:14.485037 INFO Reloading Module: /config/apps/appliances.py
2017-09-22 15:40:14.488099 INFO Loading Object dishwasher using class Dishwasher from module appliances
2017-09-22 15:40:14.489436 INFO Reloading Module: /config/apps/plants.py
2017-09-22 15:40:14.492649 INFO Loading Object fiddle_leaf_fig using class FiddleLeafFig from module plants
2017-09-22 15:40:14.493865 INFO Reloading Module: /config/apps/climate.py
2017-09-22 15:40:14.496753 INFO Loading Object air_quality using class AirQuality from module climate
2017-09-22 15:40:14.497921 INFO Reloading Module: /config/apps/alexa.py
2017-09-22 15:40:14.501088 INFO Loading Object alexa using class Alexa from module alexa
2017-09-22 15:40:14.502089 INFO App initialization complete
2017-09-22 15:40:14.539949 INFO Connected to Home Assistant 0.53.1
2017-09-22 16:00:31.128622 INFO proximity: At home
2017-09-22 18:26:44.031297 INFO living_room_lights: {'in_blackout': {UUID('7dca5394-0c63-4fc9-beef-09ce3e52db38'): {'kwargs': 1506171600, 'offset': 1506171600, 'name': 'in_blackout', 'basetime': 1506171600, 'type': None, 'repeat': 1506171600, 'interval': 1506171600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506171600}, UUID('136c2985-85a2-4d2e-b371-c9203589cd45'): {'kwargs': 1506135600, 'offset': 1506135600, 'name': 'in_blackout', 'basetime': 1506135600, 'type': None, 'repeat': 1506135600, 'interval': 1506135600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506135600}}, 'security': {UUID('6a6fe55f-8414-4339-a6f4-bb954622fa4d'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'security', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method Security.lock_up_at_midnight_cb of <security.Security object at 0x75842d70>>, 'timestamp': 1506146400}, UUID('cdb783b1-56b6-4d24-bb52-074274ed8a2b'): {'kwargs': 1506126614, 'offset': 1506126614, 'name': 'security', 'basetime': 1506126614, 'type': None, 'repeat': 1506126614, 'interval': 1506126614, 'callback': <bound method Security.absent_and_insecure_cb of <security.Security object at 0x75842d70>>, 'timestamp': 1506126614}}, 'living_room_lights': {UUID('7cc47537-5c49-40e5-bb03-4d53b81ca64c'): {'kwargs': 1506128204, 'offset': 1506128204, 'name': 'living_room_lights', 'basetime': 1506128204, 'type': 'next_setting', 'repeat': 1506128204, 'interval': 1506128204, 'callback': <bound method LivingRoomLights.sunset_cb of <lights.LivingRoomLights object at 0x75842c30>>, 'timestamp': 1506126404}}, 'systems_maintenance': {UUID('0bd212eb-b597-439e-b2f2-ea9da936aac1'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'systems_maintenance', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method SystemsMaintenance.midnight_cb of <systems.SystemsMaintenance object at 0x75842050>>, 'timestamp': 1506146400}}}
2017-09-22 18:26:44.041790 INFO living_room_lights: 30 minutes before sunset triggered; turning on lights
2017-09-22 18:28:56.032333 INFO living_room_lights: {'in_blackout': {UUID('7dca5394-0c63-4fc9-beef-09ce3e52db38'): {'kwargs': 1506171600, 'offset': 1506171600, 'name': 'in_blackout', 'basetime': 1506171600, 'type': None, 'repeat': 1506171600, 'interval': 1506171600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506171600}, UUID('136c2985-85a2-4d2e-b371-c9203589cd45'): {'kwargs': 1506135600, 'offset': 1506135600, 'name': 'in_blackout', 'basetime': 1506135600, 'type': None, 'repeat': 1506135600, 'interval': 1506135600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506135600}}, 'security': {UUID('6a6fe55f-8414-4339-a6f4-bb954622fa4d'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'security', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method Security.lock_up_at_midnight_cb of <security.Security object at 0x75842d70>>, 'timestamp': 1506146400}, UUID('cdb783b1-56b6-4d24-bb52-074274ed8a2b'): {'kwargs': 1506126614, 'offset': 1506126614, 'name': 'security', 'basetime': 1506126614, 'type': None, 'repeat': 1506126614, 'interval': 1506126614, 'callback': <bound method Security.absent_and_insecure_cb of <security.Security object at 0x75842d70>>, 'timestamp': 1506126614}}, 'living_room_lights': {UUID('7cc47537-5c49-40e5-bb03-4d53b81ca64c'): {'kwargs': 1506128204, 'offset': 1506128204, 'name': 'living_room_lights', 'basetime': 1506128204, 'type': 'next_setting', 'repeat': 1506128204, 'interval': 1506128204, 'callback': <bound method LivingRoomLights.sunset_cb of <lights.LivingRoomLights object at 0x75842c30>>, 'timestamp': 1506126392.0}}, 'systems_maintenance': {UUID('0bd212eb-b597-439e-b2f2-ea9da936aac1'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'systems_maintenance', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method SystemsMaintenance.midnight_cb of <systems.SystemsMaintenance object at 0x75842050>>, 'timestamp': 1506146400}}}
2017-09-22 18:28:56.040428 INFO living_room_lights: 30 minutes before sunset triggered; turning on lights
2017-09-22 18:43:29.033899 INFO living_room_lights: {'in_blackout': {UUID('7dca5394-0c63-4fc9-beef-09ce3e52db38'): {'kwargs': 1506171600, 'offset': 1506171600, 'name': 'in_blackout', 'basetime': 1506171600, 'type': None, 'repeat': 1506171600, 'interval': 1506171600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506171600}, UUID('136c2985-85a2-4d2e-b371-c9203589cd45'): {'kwargs': 1506135600, 'offset': 1506135600, 'name': 'in_blackout', 'basetime': 1506135600, 'type': None, 'repeat': 1506135600, 'interval': 1506135600, 'callback': <bound method InBlackout.blackout_change_cb of <hass.InBlackout object at 0x75892190>>, 'timestamp': 1506135600}}, 'security': {UUID('6a6fe55f-8414-4339-a6f4-bb954622fa4d'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'security', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method Security.lock_up_at_midnight_cb of <security.Security object at 0x75842d70>>, 'timestamp': 1506146400}, UUID('cdb783b1-56b6-4d24-bb52-074274ed8a2b'): {'kwargs': 1506127514, 'offset': 1506127514, 'name': 'security', 'basetime': 1506127514, 'type': None, 'repeat': 1506127514, 'interval': 1506127514, 'callback': <bound method Security.absent_and_insecure_cb of <security.Security object at 0x75842d70>>, 'timestamp': 1506127514}}, 'living_room_lights': {UUID('7cc47537-5c49-40e5-bb03-4d53b81ca64c'): {'kwargs': 1506128204, 'offset': 1506128204, 'name': 'living_room_lights', 'basetime': 1506128204, 'type': 'next_setting', 'repeat': 1506128204, 'interval': 1506128204, 'callback': <bound method LivingRoomLights.sunset_cb of <lights.LivingRoomLights object at 0x75842c30>>, 'timestamp': 1506126391.0}}, 'systems_maintenance': {UUID('0bd212eb-b597-439e-b2f2-ea9da936aac1'): {'kwargs': 1506146400, 'offset': 1506146400, 'name': 'systems_maintenance', 'basetime': 1506146400, 'type': None, 'repeat': 1506146400, 'interval': 1506146400, 'callback': <bound method SystemsMaintenance.midnight_cb of <systems.SystemsMaintenance object at 0x75842050>>, 'timestamp': 1506146400}}}
2017-09-22 18:43:29.044530 INFO living_room_lights: 30 minutes before sunset triggered; turning on lights
```

**Comment by bachya** [comment]
Hi there – any word on this? This is causing a bit of an issue, wherein once we decide to turn off the lights, they get turned back on automatically.

**Comment by acockburn** [comment]
I'm working on a big rewrite at the moment but I'll  look at this as soon as I am done - sorry for the delay!

**Comment by bachya** [comment]
No problem at all. Good luck!

**Comment by acockburn** [comment]
I have this replicated locally now, it only happens for negative offsets of greater than 12 minutes or so, and I didn't have any in my system :) I am doing some more troubleshooting to figure out why exactly it happens but that is some progress at least.

**Comment by bachya** [comment]
Great! Thank you for the update!

**Comment by acockburn** [comment]
I believe I have a fix for this - if it checks out tonight I'll release the fix tomorrow.

**Comment by acockburn** [comment]
Fixed in 2.1.11!

**Comment by bachya** [comment]
Awesome! Thank you so much for your diligence!
